keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,411 @@
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+  - name: worker
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+### Risks and Mitigations
+
+To avoid the JobSet become a half-baked workflow manager, we will not allow for starting particular replicated jobs.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+  
+This is an annotation added to a JobSet when `InOrder` is used.
+It will track the next ReplicatedJob that is to be started/resumed.
+
+```golang
+const (
+ ReplicatedJobStart string = "jobset.sigs.k8s.io/startup-policy-next-to-start"
+)
+
+type StartupPolicyOrderOptions string
+
+const (
+ // This is the default setting
+ // AnyOrder means that we will start jobs without any specific order.
+ AnyOrder StartupPolicyOrderOptions = "AnyOrder"
+ // InOrder starts the jobs in order that they are listed.
+ InOrder StartupPolicyOrderOptions = "InOrder"
+)
+
+type StartupPolicy struct {
+ // StartupPolicyOrder is an enum of different options for StartupPolicy
+ // AnyOrder means to start replicated jobs in any order.
+ // InOrder means to start them as they are listed in the JobSet.
+ // +kubebuilder:validation:Enum=AnyOrder;InOrder
+ StartupPolicyOrder StartupPolicyOrderOptions `json:"startupPolicyOrder"`
+}
+```
+
+### Implementation
+
+We have three cases where startupPolicy applies:
+
+- when the jobset is first created
+- when the jobset is restarted after a failure (with restart or recreate all failure policy)
+- when the jobset is resumed after suspension
+
+In cases 1 and 2, the operator controls the order of creating the jobs, and so it creates the replicatedJob only when its turn comes in the startup sequence.
+In case 3, the operator controls the order of unsuspending the jobs, and so it unsuspends the replicatedJob only when its turn comes in the startup sequence.
+
+Since each start/resume of a replicated job will trigger a reconcile, we have to track which replicated job is to start/resume.
+The annotation `jobset.sigs.k8s.io/startup-policy-next-to-start` will be updated on each reconcile to reflect what is the next ReplicatedJob to resume.
```

Comment:
Can you explain how this field will be set to prevent race conditions if the JobSet was suspended or a restart triggered mid way through a startup?

Thinking more about this, I think this is probably better be a condition on JobSet, which serves as a way to communicate startup progress.

Reply:
If its a condition, are you thinking that the reason field could reflect which replicated jobset we are waiting on?

Reply:
Yes.

Reply:
updated this.